### ISSUE TITLE
IRLBA vs RSVD

### ISSUE THREAD

**Issue by flying-sheep** [LTLA/CppIrlba#16]
Hi Aaron, idk if you remember me from years back in Cambridge.
I’m again working on scanpy! We’re currently trying to figure out discrepancies between Seurat and scanpy, with the biggest culprit being our SVD.

I haven’t gone back into the math so I wonder if you investigated what sklearn does before you adapted `irlba`. Or maybe I’m comparing apples to oranges and the two aren’t doing the same thing.

- sklearn’s default (randomized) algorithm for its [TruncatedSVD](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.TruncatedSVD.html) is based on [Halko, et al. (2009)](https://arxiv.org/abs/0909.4061):

  > [It] performs linear dimensionality reduction[.] Contrary to PCA, this estimator does not center the data before computing the [SVD]. This means it can work with sparse matrices efficiently

- irlba is based on [Baglama & Reichel (2005)](https://www.researchgate.net/publication/220411989_Augmented_Implicitly_Restarted_Lanczos_Bidiagonalization_Methods):

  > [It defines methods for] computation of a few of the largest or smallest singular values of a large [sparse] matrix

If they do comparable things, I wonder if anybody has ever compared their accuracy, speed, and so on to each other.

cc @ilan-gold

PS: marginally relevant: https://bwlewis.github.io/irlba/comparison.html but I don’t think RSpectra/ARPACK is that relevant, as `TruncatedSVD` doesn’t default to its ARPACK implementation.

**Comment by LTLA** [comment]
Yes, I do recall comparing those two algorithms; both are wrapped by **BiocSingular**, with the equivalent of sklearn's truncated SVD being implemented in `rsvd::rsvd()` IIUC. And yes, `rsvd()` gives different results from `irlba()`, and I believe this is expected from theory (based on some conversations I had with **rsvd** author). To paraphrase, RSVD is less accurate but this should be acceptable provided that the singular values decay quickly.

In practice, I recall two major differences between the two algorithms:

- RSVD was more sensitive to the choice of parameters. I don't remember exactly, but I think it was sensitive to the seed and to the target rank. Most of this "instability" occurs at later singular vectors, presumably where the singular value decay is slower and thus any changes in accuracy are more pronounced. IRLBA was more robust to its parameter choices, which inspires a bit more confidence in its reliability (accuracy notwithstanding).
- RSVD is much faster for file-backed matrices as it involves fewer scans through the matrix. So for a while, we were using it as the default when processing any file-backed matrix. But in the end, I got tired of dealing of my HPC's slow I/O and decided to just load everything into memory and analyze it there, at which point I defaulted to IRLBA. (I don't think, in my hands, RSVD was much faster than IRLBA for in-memory matrices; even though there are fewer matrix multiplications in RSVD, each multiplication involves more column vectors, so it more or less cancels out.)

That said, these memories are almost a decade old, so YMMV.

(I won't bother him directly here, but I'll just mention erichson/rSVD#4, which is where I started talking to the **rsvd** author, just in case you want to talk to someone who actually knows the math.)

**Comment by flying-sheep** [comment]
Thanks, much appreciated! In Python-land we’re doing a lot of [Dask](https://www.dask.org/)-based file-backed computing these days, which means that RSVD is probably the better choice for these use cases.

But maybe it’s time someone implements fast IRLBA for Python as an alternative, stability can be nice!

(I’m personally not a huge fan of relying on it: if one’s results are dependent on some algorithm being stable, they probably don’t have the statistical legs to stand on)
